README.md
## vegperiod: Determine Thermal Vegetation Periods 

<!-- badges: start -->
[![R-CMD-check](https://github.com/rnuske/vegperiod/actions/workflows/R-CMD-check.yaml/badge.svg)](https://github.com/rnuske/vegperiod/actions/workflows/R-CMD-check.yaml)
[![Package-License](https://img.shields.io/badge/license-GPL--3-brightgreen.svg?style=flat)](https://www.gnu.org/licenses/gpl-3.0.html) 
[![CRAN](https://www.r-pkg.org/badges/version/vegperiod)](https://cran.r-project.org/package=vegperiod) 
[![Dependencies](https://tinyverse.netlify.com/badge/vegperiod)](https://cran.r-project.org/package=vegperiod) 
[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.1466541.svg)](https://doi.org/10.5281/zenodo.1466541)
<!-- badges: end -->

The vegetation period, or growing season, is the period of the year when the weather conditions are sufficient for plants to grow. This package provides methods to calculate climatological or thermal growing seasons solely based on daily mean temperatures and the day of the year (DOY). Because of their simplicity, they are commonly used in plant growth models and climate change impact assessments.

The concept of a temperature driven vegetation period holds mostly for the temperate climate zone. At lower latitudes, other factors such as precipitation and evaporation can be more decisive. Some methods such as GSL of `ETCCDI` are employed globally (with a half year shift in the southern hemisphere). Others have a smaller area of application as they have been parameterized with local to regional observations. However, the methods `Menzel` and `vonWilpert` are used throughout Germany.

The package also includes functions for downloading open meteo data from Germany's National Meteorological Service (Deutscher Wetterdienst, DWD).


### Installation
The stable version can be installed from CRAN
```r
install.packages("vegperiod")
```

and the development version is available from Github using the package `remotes`
```r
remotes::install_github("rnuske/vegperiod")
```

### Usage
Vegetation periods are calculated using the function `vegperiod()`.  One has to choose at least a start and an end method. Some methods require additional arguments, such as 'Menzel' which needs 'species'.

```r
data(goe)
vegperiod(dates=goe$date, Tavg=goe$t, 
          start.method="Menzel", end.method="vonWilpert", 
          species="Picea abies (frueh)", est.prev=3)
```

### Implemented start and end methods
Common methods for determining the onset and end of thermal vegetation periods are provided, for details see next sections and documentation. Suggestions or contributions of additional methods are always welcome. Popular choices with regard to forest trees in Germany are `Menzel` and `vonWilpert`. Climate change impact studies at NW-FVA are frequently conducted using `Menzel` with "Picea abies (frueh)" and `NuskeAlbert` for all tree species; with tree species specifics accounted for in subsequent statistical models.

#### Start methods
* **Menzel** implemented as described in Menzel (1997). Parameterized for 10 common tree species. Requires previous years chill days, which can be substituted by the average of first years.
* **StdMeteo** / **ETCCDI** a simple threshold based procedure as defined by the Expert Team on Climate Change Detection and Indices (cf. ETCCDI 2009 and Frich et al. 2002). Leading to quite early vegetation starts.
* **Ribes uva-crispa** using leaf out of gooseberry as indicator. Developed by Germany's National Meteorological Service (Deutscher Wetterdienst, DWD). Presented in the section forestry of DWD's 'Climate Atlas'. It is more robust against early starts than common simple meteorological procedures.

#### End methods
* **vonWilpert** based on von Wilpert (1990). Originally developed for "Picea abies" in the Black Forest but currently used for all tree species throughout Germany.
* **LWF-BROOK90** a reimplementation of the LWF-BROOK90 VBA version of vonWilpert (Hammel and Kennel 2001).
* **NuskeAlbert** a very simple method inspired by standard climatological practices.
* **StdMeteo** / **ETCCDI** a simple threshold based procedure as defined by the Expert Team on Climate Change Detection and Indices (cf. ETCCDI 2009 and Frich et al., 2002). Leading to quite late vegetation ends.

### Download climate data from DWD (German Meteorological Service)
Germany's National Meteorological Service offers open meteo data in its [Climate Data Center](https://cdc.dwd.de/portal/).
The files are organized in deep folder structures and end with an arcane/legacy EOF character. 
The Function `read.DWDdata()`deals with all of that and returns a `data.frame`. Beware there might be missing values and inhomogeneities.

Note: Downloading 'historical' data from DWD with `read.DWDdata()` requires the package 'curl'.


### Contributions
Implementations of further start and end methods or download functions are more than welcome! Please suggest suitable candidates via issue or pull request.


NEWS.md
# News for Package vegperiod

## Version 0.4.0
  * introduced threshold `Tsum.crit` for sum of day degrees in function `vegperiod()`
  * made plausibility check optional via new parameter `check.data` (suggested by @rhabel)
  * restructured documentation to provide more details on start and end methods 
  * acknowledged funder in DESCRIPTION
  * fixed a changed URL


## Version 0.3.1
  * if `StdMeteo` does not find an end, it now reports the last day of the year instead of 0.


## Version 0.3
  * fix `est.prev` parameter used in start method Menzel. Now averages as advertised (Thanks for catching this bug to @awellpott).
  * switched to unit testing framework `tinytest`


## Version 0.2.7
  * changed temperature plausibility checks because of severe winter 1956 with daily mean temperatures of -27 °C in Germany and even lower and higher daily mean temperatures in climate scenarios.


## Version 0.2.6
  * fixed change of base URL of DWD open data (Thanks @ckluss for reporting and pointing in the right direction)
  * fixed download of recent monthly climate data. Breakage caused by DWD file name schema change (#1). Thanks @rwoerde for reporting.
  * corrected minimum R version. Argument `fileEncoding` was introduced to `read.fwf()` in R 3.2


## Version 0.2.5
  * replaced package 'RCurl' with 'curl' for downloading DWD Data
  * stopped importing package 'methods'
  * fixed read.DWDstations() because DWD changed layout of station list
  * updated documentation of read.DWDdata() because DWD changed column order of climate data files


## Version 0.2.4
  * vegperiod() checks if Tavg (daily mean temperature) is between -25 and 35 to catch temperatures multiplied by 10. Sometimes done to store as integers. (thanks to Ronald Bialozyt for reporting a crash)


## Version 0.2.3
  * vegperiod() reports end of vegetation period always as integer
  * new parameter 'quiet' to readDWDdata to suppress messages and progessbar of download
  * started unit testing with package 'testthat'


## Version 0.2.2
  * start.method='Ribes uva-crispa' using leaf out of gooseberry as indicator. Developed by Germany's National Meteorological Service (DWD).
  * est.prev can now be equal to number of provided years
  * proper error message if only one year of data provided and previous year's chill days requested


## Version 0.2.1
  * cleaned up mangled output of start.method='StdMeteo'. StdMeteo now behaves properly and returns a vector.
  * use stats::filter() in the package to protected filter() against getting run over by package dplyr (thanks to Cristabel Duran for reporting).


## Version 0.2.0
### New Features
  * `read.DWDdata()` reads open meteo data from the Climate Data Center (CDC) of Germany's National Meteorological Service (Deutscher Wetterdienst, DWD).
  * `read.DWDstations()` fetches information about available DWD stations

### Major Changes
  * renamed arguments in vegperiod() to clarify meaning / easier tab completion:
    * first.avg -> est.prev 
    * temp.sum -> Tsum.out
    * method.start -> start.method
    * method.end -> end.method
  * renamed method `StdClimatolgy` to `StdMeteo` / `ETCCDI` in vegperiod() and brought calculations into compliance with ETCCDI standard

### Bug fixes and minor improvements
  * added a hack in method LWF-BROOK90 of vegperiod() to adjust for different arithmetic precision in VBA version (thanks to Paul Schmidt-Walter for providing climate.in files for comparision and help finding the bug)
  * fixed bug leading to crash of vegperiod() if argument `first.avg=0` (thanks to Ronald Bialozyt for reporting)
  * the argument "dates" of vegperiod() accepts now also character strings understood by `as.Date()`
  * rewrote help-page of vegperiod() and added references for all methods


## Version 0.1.0
  * Intial version


R/vegperiod.R
#==============================================================================
# Vegetation Period
#   calculate start and end of vegetation periode
#
# rewrite 2015-02-15 - 2015-02-26  by Robert Nuske
#  Written on my way to work.
#  Thanks to good working conditions (no internet) in the ICE.
#==============================================================================

#' Determine vegetation period
#'
#' Calculate start and end date of vegetation periods based on daily average
#' air temperature and the day of the year (DOY).
#' The sum of day degrees within the vegetation period is included for convenience.
#'
#'
#' @details
#' Common methods for determining the onset and end of thermal vegetation
#' periods are provided, for details see next sections. Popular choices with
#' regard to forest trees in Germany are `Menzel` and `vonWilpert`. Climate
#' change impact studies at NW-FVA are frequently conducted using `Menzel` with
#' "Picea abies (frueh)" and `NuskeAlbert` for all tree species; with tree
#' species specifics accounted for in subsequent statistical models.
#'
#' ## Start methods:
#' The method **[`Menzel`][method_Menzel]** implements the algorithm described in
#' Menzel (1997). The method is parameterized for 10 common tree species. It
#' needs previous year's chill days. **[`ETCCDI`][method_ETCCDI]** resp.
#' `StdMeteo` is a simple threshold based procedure as defined by the
#' Expert Team on Climate Change Detection and Indices (cf. ETCCDI 2009, Frich
#' et al. 2002, Zhang et al. 2011) leading to quite early vegetation starts.
#' This method is widely used in climate change studies. The method
#' **[`Ribes uva-crispa`][method_Ribes]** is based on leaf-out of gooseberry (Janssen
#' 2009). It was developed by the Germany's National Meteorological Service
#' (Deutscher Wetterdienst, DWD) and is more robust against early starts than
#' common simple meteorological procedures.
#'
#' ## End methods:
#' The end method **[`vonWilpert`][method_vonWilpert]** is based on von Wilpert (1990). It
#' was originally developed for *Picea abies* in the Black Forest but is
#' commonly used for all tree species throughout Germany. As usual, the rules
#' regarding the soilmatrix are neglected in this implementation.
#' **[`LWF-BROOK90`][method_LWF-BROOK90]** is -for the sake of convenience- a
#' reimplementation of the LWF-BROOK90 VBA (version 3.4) variant of "vonWilpert"
#' (Hammel and Kennel 2001). Their interpretation of von Wilpert (1990) and the
#' somewhat lower precision of VBA was mimicked. **[`NuskeAlbert`][method_NuskeAlbert]**
#' provide a very simple method which is inspired by standard climatological
#' procedures but employs a 7 day moving average and a 5 °C threshold (cf.
#' Walther and Linderholm 2006). **[`ETCCDI`][method_ETCCDI]** resp. `StdMeteo`
#' is a simple threshold based procedure as defined by the Expert Team on
#' Climate Change Detection and Indices (cf. ETCCDI 2009, Frich et al. 2002,
#' Zhang et al. 2011) leading to quite late vegetation ends.
#'
#' @param dates vector of calendar dates (objects of class `Date` or something
#'   understood by [as.Date()]). Must contain entire years if `est.prev > 0`
#'   else the first year may comprise only November and December.
#' @param Tavg vector of daily average air temperatures in degree Celsius.
#'   Same length as `dates`.
#' @param start.method name of method to use for vegetation start. One of
#'   `"Menzel"` (needs additional argument `species`, see below), `"StdMeteo"`
#'   resp. `"ETCCDI"`, `"Ribes uva-crispa"`. Can be abbreviated (partial
#'   matching). For further discussion see Details.
#' @param end.method name of method to use for vegetation end. One of
#'   `"vonWilpert"`, `"LWF-BROOK90"`, `"NuskeAlbert"` and `"StdMeteo"` resp.
#'   `"ETCCDI"`. Can be abbreviated (partial matching). For further discussion
#'   see Details.
#' @param Tsum.out boolean. Return the sum of daily mean temperatures above
#'   `Tsum.crit` within vegetation period, also known as growing day degrees.
#' @param Tsum.crit threshold for sum of day degrees. Only daily mean temperatures
#'   `> Tsum.crit` will be tallied. The default of `0` prevents negative
#'   daily temperatures from reducing the sum. Climate change studies often use
#'   a threshold of `5`.
#' @param species name of tree species (required if `start.method="Menzel"`
#'   ignored otherwise).
#'
#'   Must be one of `"Larix decidua"`, `"Picea abies (frueh)"`,
#'   `"Picea abies (spaet)"`, `"Picea abies (noerdl.)"`, `"Picea omorika"`,
#'   `"Pinus sylvestris"`, `"Betula pubescens"`, `"Quercus robur"`,
#'   `"Quercus petraea"`, `"Fagus sylvatica"`.
#' @param est.prev number of years to **est**imate **prev**ious year's chill
#'   days for the first year (required if `start.method="Menzel"` ignored
#'   otherwise).
#'
#'   `Menzel` requires the number of chill days of previous November and
#'   December. If `est.prev = 0` the first year is used to get the previous
#'   year's chill days and dropped afterwards. Thus, a year from the time
#'   series is lost. To avoid losing a year, `est.prev = n` estimates the
#'   previous year's chill days for the first year from the average of `n`
#'   first years of the time series.
#' @param check.data Performs plausibility checks on the temperature data to
#'   ensure that the temperatures have not been multiplied by ten.
#'   Plausible range is -35 to +40°C.
#'
#' @return A data.frame with year and DOY of start and end day of
#'   vegetation period. If `Tsum.out=TRUE`, the data.frame contains an
#'   additional column with the sum of day degrees within vegetation periods.
#'
#' @references
#'   ETCCDI (2009)
#'   Climate Change Indices: Definitions of the 27 core indices.
#'   \url{http://etccdi.pacificclimate.org/list_27_indices.shtml}
#'
#'   Frich, P., Alexander, L., Della-Marta, P., Gleason, B., Haylock, M.,
#'   Klein Tank, A. and Peterson, T. (2002)
#'   Observed coherent changes in climatic extremes during the second half of
#'   the twentieth century.
#'   \emph{Climate Research}, \bold{19}, 193--212.
#'   \doi{10.3354/cr019193}.
#'
#'   Hammel, K. and Kennel, M. (2001)
#'   Charakterisierung und Analyse der Wasserverfügbarkeit und des
#'   Wasserhaushalts von Waldstandorten in Bayern mit dem Simulationsmodell
#'   BROOK90.
#'   \emph{Forstliche Forschungsberichte München}.
#'
#'   Janssen, W. (2009)
#'   Definition des Vegetationsanfanges.
#'   \emph{Internal Report, Deutscher Wetterdienst, Abteilung Agrarmeteorologie}.
#'
#'   Menzel, A. (1997)
#'   Phänologie von Waldbäumen unter sich ändernden Klimabedingungen -
#'   Auswertung der Beobachtungen in den Internationalen Phänologischen Gärten
#'   und Möglichkeiten der Modellierung von Phänodaten.
#'   \emph{Forstliche Forschungsberichte München}.
#'
#'   von Wilpert, K. (1990)
#'   Die Jahrringstruktur von Fichten in Abhängigkeit vom Bodenwasserhaushalt
#'   auf Pseudogley und Parabraunerde: Ein Methodenkonzept zur Erfassung
#'   standortsspezifischer Wasserstreßdispostion.
#'   \emph{Freiburger Bodenkundliche Abhandlungen}.
#'
#'   Walther, A. and Linderholm, H. W. (2006)
#'   A comparison of growing season indices for the Greater Baltic Area.
#'   \emph{International Journal of Biometeorology}, \bold{51}(2), 107--118.
#'   \doi{10.1007/s00484-006-0048-5}.
#'
#'   Zhang, X., Alexander, L., Hegerl, G. C., Jones, P., Tank, A. K.,
#'   Peterson, T. C., Trewin, B. and Zwiers, F. W. (2011)
#'   Indices for monitoring changes in extremes based on daily temperature and
#'   precipitation data.
#'   \emph{Wiley Interdisciplinary Reviews: Climate Change}, \bold{2}(6), 851--870.
#'   \doi{10.1002/wcc.147}.
#'
#' @examples
#' data(goe)
#' vegperiod(dates=goe$date, Tavg=goe$t,
#'           start.method="Menzel", end.method="vonWilpert",
#'           species="Picea abies (frueh)", est.prev=5)
#'
#' # take chill days from first year, which is then dropped
#' vegperiod(dates=goe$date, Tavg=goe$t, start="Menzel", end="vonWilpert",
#'           species="Picea abies (frueh)", est.prev=0)
#'
#' # add column with sum of day degrees in vegetation periods
#' vegperiod(dates=goe$date, Tavg=goe$t, Tsum.out=TRUE,
#'           start="StdMeteo", end="StdMeteo")
#' @md
#' @export
vegperiod <- function(dates, Tavg, start.method, end.method, Tsum.out=FALSE,
                      Tsum.crit=0, species=NULL, est.prev=0, check.data=TRUE){
  # Checks
  #----------------------------------------------------------------------------
  start.method <- match.arg(start.method,
                            choices=c('Menzel', 'StdMeteo', 'ETCCDI',
                                      'Ribes uva-crispa'))
  end.method <- match.arg(end.method,
                          choices=c('vonWilpert', 'LWF-BROOK90', 'NuskeAlbert',
                                    'StdMeteo', 'ETCCDI'))

  if(start.method == 'Menzel'){
    possible.species <- c("Larix decidua", "Picea abies (frueh)",
                          "Picea abies (spaet)", "Picea abies (noerdl.)",
                          "Picea omorika", "Pinus sylvestris",
                          "Betula pubescens", "Quercus robur",
                          "Quercus petraea", "Fagus sylvatica")
    if(is.null(species))
      stop(paste0("If start.method='Menzel', species must be one of '",
                  paste(possible.species, collapse="', '"), "'."))

    species <- match.arg(species, choices=possible.species)
  }

  if(length(dates) != length(Tavg))
    stop("The arguments dates and Tavg must be of same length!")

  # are the temperatures sound
  if(isTRUE(check.data)){
  minimax <- range(Tavg)
  if(minimax[1] < -35 | minimax[2] > 40)
    stop("Your input temperature data exceeds the plausible range of -35 to +40\u00b0C.\n",
         "You may want to double check your data. ",
         "If you still want to use the given data, set 'check.data' to FALSE")
  }

  if(!inherits(dates, "Date")){
    tryCatch(dates <- as.Date(dates),
             error=function(c) stop(paste("'dates' could not be coerced to",
                                          "class Date by as.Date(dates)."),
                                    call.=FALSE)
    )
  }

  # set up the workhorse data.frame
  df <- data.frame(year  = as.integer(format(dates, "%Y")),
                   month = as.integer(format(dates, "%m")),
                   DOY   = as.integer(format(dates, "%j")),
                   Tavg)
  rm(dates, Tavg)

  # determine leap years
  years <- unique(df$year)
  leap <- (years %% 4 == 0 & years %% 100 != 0) | years %% 400 == 0

  # est.prev in valid range?
  # set to a high value if not Menzel
  if(start.method == 'Menzel'){
    est.prev <- as.integer(est.prev)
    if(est.prev < 0 || est.prev > length(years))
      stop("est.prev must be >= 0 and <= number of years.")
    if(est.prev == 0 && length(years) == 1)
      stop(paste("It's not possible to use previous year's chill days",
           "(est.prev=0) if only one year is provided.",
           "Either set est.prev=1 or provide more years."))
  } else {
    est.prev <- 999
  }

  # Do the provided years have the correct length (considering leap years)?
  # last year only till October 5th (DOY=279) needed
  if(est.prev == 0){
    # Menzel: first year discarded -> only Nov+Dec of first year needed
    year.length <- c(61, 365 + leap[c(-1, -length(leap))], 279)
  } else {
    # full first year required
    # either not Menzel or guesstimate first year's NovDecCD from previous yrs
    year.length <- c(365 + leap[-length(leap)], 279)
  }
  if(!all(tapply(df$DOY, df$year, FUN=length) >= year.length))
    stop("Not all years in 'dates' have expected length.")

  if(est.prev == 0 &&
     length(df[df$year == years[1] & df$month >= 11, "DOY"]) != 61)
    stop("First year must at least contain full November & December")


  # calculating start and end
  #----------------------------------------------------------------------------
  start <- switch(start.method,
                  'Menzel'   = .start_menzel(df=df, est.prev=est.prev, species=species),
                  'StdMeteo' = ,
                  'ETCCDI'   = .start_std_meteo(df),
                  'Ribes uva-crispa' = .start_ribes(df))

  # if first year only used for getting previous cold days -> drop it now
  if(est.prev == 0){
    df <- df[df$year != years[1], ]
    years <- years[-1]
  }

  end <- switch(end.method,
                'vonWilpert'  = .end_vonWilpert(df),
                'LWF-BROOK90' = .end_LWF_BROOK90(df),
                'NuskeAlbert' = .end_NuskeAlbert(df, start),
                'StdMeteo'    = ,
                'ETCCDI'      = .end_std_meteo(df))


  # collect results
  #----------------------------------------------------------------------------
  # if indicated calculate day degrees in vegperiod also
  res <- data.frame(year=years, start=start, end=end)

  seq_len

  if(Tsum.out){
    res$Tsum <- numeric(nrow(res))
    for(i in seq_along(res$Tsum)){
      res$Tsum[i] <- sum(df$Tavg[df$year == years[i] &
                                 df$DOY  >= start[i] &
                                 df$DOY  <= end[i]   &
                                 df$Tavg >  Tsum.crit])
    }
  }

  return(res)
}
